data_structures/linked_list.ts

```diff
@@ -0,0 +1,475 @@
+   * assertEquals([...list], [1, 2, 3, 4, 5, 6, 7, 8]);
+   * ```
+   */
+  insert(value: T, index: number): void {
+    if (index === 0) {
+      return this.unshift(value);
+    }
+
+    if (index === this.#length) {
+      return this.push(value);
+    }
+
+    const dt = this.#length - index - 1;
+
+    if (index < 0 || dt < 0) {
+      throw new RangeError(
+        "Cannot insert the value: The index is out of range",
+      );
+    }
+
+    const node: LinkedNode<T> = { value, prev: null, next: null };
+    let ptr: LinkedNode<T>;
+
+    if (dt < this.#length / 2) {
+      ptr = this.#tail!;
+      for (let i = 0; i < dt; ++i) {
+        ptr = ptr.prev!;
+      }
+    } else {
+      ptr = this.#head!;
+      for (let i = 0; i < index; ++i) {
+        ptr = ptr.next!;
+      }
+    }
+
+    node.next = ptr;
+    node.prev = ptr.prev;
+    ptr.prev!.next = node;
+    ptr.prev = node;
+
+    ++this.#length;
+  }
+
+  /**
+   * Removes the node at the given index and returns its value, if any.
+   *
+   * The complexity of this operation is linear O(n).
+   *
+   * @param index The position to remove.
+   * @returns The removed node's value, `undefined` is returned if:
+   * - `length == 0`
+   * - `index < 0`
+   * - `index >= length`
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from([2, 4, 6, 8]);
+   * assertEquals([...list], [2, 4, 6, 8]);
+   *
+   * list.remove(1);
+   * assertEquals([...list], [2, 6, 8]);
+   *
+   * list.remove(2);
+   * assertEquals([...list], [2, 6]);
+   * ```
+   */
+  remove(index: number): T | undefined {
+    if (index === 0) {
+      return this.shift();
+    }
+
+    if (index === this.#length - 1) {
+      return this.pop();
+    }
+
+    const dt = this.#length - index - 1;
+
+    if (index < 0 || dt < 0) {
+      return undefined;
+    }
+
+    let ptr: LinkedNode<T>;
+
+    if (dt < this.#length / 2) {
+      ptr = this.#tail!;
+      for (let i = 0; i < dt; ++i) {
+        ptr = ptr.prev!;
+      }
+    } else {
+      ptr = this.#head!;
+      for (let i = 0; i < index; ++i) {
+        ptr = ptr.next!;
+      }
+    }
+
+    const value = ptr.value;
+    ptr.next!.prev = ptr.prev;
+    ptr.prev!.next = ptr.next;
+
+    --this.#length;
+    return value;
+  }
+
+  /**
+   * Determines whether the linked list contains a node whose value
+   * equals to the given value.
+   *
+   * The complexity of this operation is linear O(n).
+   *
+   * @param value The value to check
+   * @returns `true` if the list contains the given value.
+   *
+   * @example Usage
+   * ```ts
+   * import { LinkedList } from "@std/data-structures";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const list = LinkedList.from([2, 4, 6, 8]);
+   *
+   * assertEquals(list.has(4), true);
+   * assertEquals(list.has(0), false);
+   * ```
+   */
+  has(value: T): boolean {
```

Comment:
If we want to follow the array convention this probably should be called `includes(value: T)`